### PR TITLE
CLN: reshape

### PR DESCRIPTION
**Pull request by jbrockmendel**
I think some of this can be made easier to reason about (like with making FrameApply less stateful), as a preliminary this is just some mild annotations etc

### CODE REVIEW

File: pandas/core/reshape/concat.py [hunk 1]
```diff
@@ -437,13 +437,13 @@ def get_result(self):
...
                 )
-                cons = _get_series_result_type(mgr, self.objs)
+                cons = self.objs[0]._constructor
                 return cons(mgr, name=name).__finalize__(self, method="concat")
 
             # combine as columns in a frame
             else:
                 data = dict(zip(range(len(self.objs)), self.objs))
-                cons = _get_series_result_type(data)
+                cons = DataFrame
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: does this play nice with subclassing?

File: pandas/core/reshape/melt.py [hunk 2]
```diff
@@ -129,6 +131,8 @@ def lreshape(data, groups, dropna=True, label=None):
     groups : dict
         {new_name : list_of_columns}
     dropna : boolean, default True
+    label : object, default None
```
Comment (jreback) [inline pandas/core/reshape/melt.py]: should deprecate / remove this

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -1299,19 +1299,21 @@ def _get_join_indexers(
         right_keys
     ), "left_key and right_keys must be the same length"
 
-    # bind `sort` arg. of _factorize_keys
-    fkeys = partial(_factorize_keys, sort=sort)
-
     # get left & right join labels and num. of levels at each location
-    llab, rlab, shape = map(list, zip(*map(fkeys, left_keys, right_keys)))
+    mapped = (
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: can use enumerate here instead of range(len(...))

File: pandas/core/reshape/merge.py [hunk 4]
```diff
@@ -1775,11 +1777,11 @@ def flip(xs):
...
 def _get_multiindex_indexer(join_keys, index: MultiIndex, sort: bool):
 
-    # bind `sort` argument
-    fkeys = partial(_factorize_keys, sort=sort)
-
     # left & right join labels and num. of levels at each location
-    mapped = (fkeys(index.levels[n], join_keys[n]) for n in range(len(index.levels)))
+    mapped = (
+        _factorize_keys(index.levels[n], join_keys[n], sort=sort)
+        for n in range(len(index.levels))
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: use enumerate

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/concat.py [hunk 5]
```diff
@@ -541,7 +537,7 @@ def _get_concat_axis(self):
                 idx = ibase.default_index(len(self.objs))
                 return idx
             elif self.keys is None:
-                names = [None] * len(self.objs)
+                names = [None] * len(self.objs)  # type: list
```
Comment (WillAyd) [inline pandas/core/reshape/concat.py]: At least for now should always prefer `typing.*` to the builtin containers

File: pandas/core/reshape/melt.py [hunk 6]
```diff
@@ -412,14 +416,14 @@ def wide_to_long(df, stubnames, i, j, sep: str = "", suffix: str = r"\d+"):
                 two  2.9
     """
 
-    def get_var_names(df, stub, sep, suffix):
+    def get_var_names(df, stub: str, sep: str, suffix) -> List[str]:
```
Comment (WillAyd) [inline pandas/core/reshape/melt.py]: I think can annotation suffix as a str

File: pandas/core/reshape/concat.py [hunk 7]
```diff
@@ -537,7 +538,7 @@ def _get_concat_axis(self):
                 idx = ibase.default_index(len(self.objs))
                 return idx
             elif self.keys is None:
-                names = [None] * len(self.objs)
+                names = [None] * len(self.objs)  # type: List
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: can you new typing format

File: pandas/core/reshape/melt.py (hunk 2 above)
Comment (jreback) [reply pandas/core/reshape/melt.py]: can you create an issue

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/concat.py (hunk 1 above)
Comment (jreback) [reply pandas/core/reshape/concat.py]: ?

Comment (jbrockmendel) [reply pandas/core/reshape/concat.py]: probably not, but its the exact behavior we currently have

File: pandas/core/reshape/melt.py (hunk 2 above)
Comment (jreback) [reply pandas/core/reshape/melt.py]: can you create an issue to deprecate this

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, if you can create an issue for that deprecation

**Pull request merged**